lib/view_component/inline_template.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module ViewComponent # :nodoc:
+  module InlineTemplate
+    extend ActiveSupport::Concern
+    Template = Struct.new(:source, :language, :path, :lineno)
+
+    class_methods do
+      def template!(template)
+        caller = caller_locations(1..1)[0]
```

Comment:
This makes a small assumption, which is that all `template!` calls will be formatted like:

```ruby
template <<~ERB
ERB
```

and not

```ruby
template!(
  <<~ERB
  ERB
)
```

The newline is what's problematic here, since we assume that the template will start on the same line as the method call (`template!`). I think that's fine, but if it becomes a problem we can provide more "fine grained" options like allowing an offset to be passed

Reply:
It also cannot be:

```ruby
template! "<div>Alert!"
tmplt = "asdf"
template! tmplt
def self.generate_template
  "a_template"
end
template! generate_template
template! File.read("some_file.html.erb")
```

Reply:
I'm not a fan of imposing this constraint. We should accept all valid Ruby.

Reply:
What if we were to wrap template errors with a message along the lines of "Error on line 3 of MyComponent inline template"

Reply:
I have some ideas here on how we might be able to improve the error messages, but I'd prefer they weren't blocking. Any opposition to merging this as-is and exploring better error strategies later? This is marked as experimental, so a little bit of churn seems fine here imo.